File: matlab/src/opt/+opt/ThroughputConstraint.m
classdef ThroughputConstraint < opt.Constraint
    % ThroughputConstraint  Throughput constraint: Tput >= minValue.
    properties
        station
        jobclass = '';
        minValue
    end
    methods
        function obj = ThroughputConstraint(station, jobclass, minValue, name)
            if nargin < 4, name = ''; end
            obj@opt.Constraint(name);
            obj.station = opt.Constraint.nameOf(station);
            if nargin >= 2 && ~isempty(jobclass)
                obj.jobclass = opt.Constraint.nameOf(jobclass);
            end
            obj.minValue = minValue;
        end
        function n = generateName(obj)
            if ~isempty(obj.jobclass)
                n = sprintf('Tput_%s_%s_ge_%g', obj.station, obj.jobclass, obj.minValue);
            else
                n = sprintf('Tput_%s_ge_%g', obj.station, obj.minValue);
            end
        end
        function v = evaluate(obj, result, ~)
            if ~isempty(obj.jobclass)
                actual = result.getThroughput(obj.station, obj.jobclass);
            else
                actual = result.getThroughput(obj.station);
            end
            v = opt.Constraint.lowerBoundViolation(actual, obj.minValue);
        end
    end
end

File: jar/src/main/java/jline/opt/objectives/ThroughputConstraint.java
package jline.opt.objectives;

import jline.lang.JobClass;
import jline.lang.nodes.Station;
import jline.opt.results.EvaluationResult;

import java.util.Map;

/** Throughput constraint: Tput &gt;= minValue. */
public class ThroughputConstraint extends Constraint {

    private final String station;
    private final String jobclass;   // null for aggregate
    private final double minValue;

    public ThroughputConstraint(String station, String jobclass, double minValue, String name) {
        super(name);
        this.station = station;
        this.jobclass = jobclass;
        this.minValue = minValue;
    }

    public ThroughputConstraint(Station station, JobClass jobclass, double minValue) {
        this(station.getName(), jobclass == null ? null : jobclass.getName(), minValue, null);
    }

    public ThroughputConstraint(Station station, double minValue) {
        this(station.getName(), null, minValue, null);
    }

    protected String generateName() {
        if (jobclass != null) {
            return "Tput_" + station + "_" + jobclass + "_ge_" + minValue;
        }
        return "Tput_" + station + "_ge_" + minValue;
    }

    public String getStation() {
        return station;
    }

    public double getMinValue() {
        return minValue;
    }

    public double evaluate(EvaluationResult result, Map<String, Object> variableValues) {
        double actual = (jobclass != null)
                ? result.getThroughput(station, jobclass)
                : result.getThroughput(station);
        return lowerBoundViolation(actual, minValue);
    }
}
